src/App.tsx

```diff
@@ -1,223 +1,140 @@
-/* eslint-disable jsx-a11y/control-has-associated-label */
-import React, { useContext, useEffect, useRef } from 'react';
+import React, {
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react';
+import { createTodo, deleteTodo, getTodos } from './api/todos';
 import { AuthContext } from './components/Auth/AuthContext';
+import { ErrorNotification }
+  from './components/ErrorNotification/ErrorNotification';
+import { FilterTodos } from './components/FilterTodos/FilterTodos';
+import { NewTodo } from './components/NewTodo/NewTodo';
+import { TodoList } from './components/TodoList/TodoList';
+import { Errors } from './types/Errors';
+import { GroupBy } from './types/GroupBy';
+import { Todo } from './types/Todo';
 
 export const App: React.FC = () => {
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
   const user = useContext(AuthContext);
   const newTodoField = useRef<HTMLInputElement>(null);
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [visibleTodos, setVisibleTodos] = useState(todos);
+  const [error, setError] = useState<Errors>(Errors.NONE);
+  const [isAdding, setIsAdding] = useState(false);
+  const [isDeleting, setIsDeleting] = useState(false);
+
+  const loadTodos = async () => {
+    try {
+      setTodos(await getTodos(user?.id || 0));
+    } catch {
+      setError(Errors.URL);
+    }
+  };
 
   useEffect(() => {
-    // focus the element with `ref={newTodoField}`
     if (newTodoField.current) {
       newTodoField.current.focus();
     }
-  }, []);
+
+    loadTodos();
+  }, [isAdding]);
+
+  const filterTodos = useCallback((groupBy: string) => {
+    setVisibleTodos(
+      todos.filter(todo => {
+        switch (groupBy) {
+          case GroupBy.Active:
+            return !todo.completed;
+          case GroupBy.Completed:
+            return todo.completed;
+
+          default:
+            return true;
+        }
+      }),
+    );
+  }, [todos]);
+
+  const completedTodos = useMemo(() => (
+    todos.filter(({ completed }) => completed)
+  ), [todos]);
+
+  const leftTodos = useMemo(() => (
+    todos.filter(({ completed }) => !completed)
+  ), [todos]);
+
+  const addTodo = async (todoTitle: string) => {
+    setIsAdding(true);
+
+    try {
+      setTodos([...todos, await createTodo(user?.id || 0, todoTitle)]);
```

Comment:
If the new state is computed using the previous state, you can pass a function to setTodos